File: src/engine/customEvents.ts
import { POST_WINDOW_TD, PRE_WINDOW_TD, TRIGGER_ASSET } from '@/config/engine';
import { DailyHistoryPayload } from '@/engine/types';
import { AssetMeta } from '@/engine/returns';

export interface HistoricalPoint {
  index: number;
  date: string;
  value: number;
}

export interface CustomEventComputation {
  returnsByAsset: Record<string, Record<number, number>>;
  selectedDate: string;
  resolvedAnchorDate: string | null;
  coverage: {
    startDate: string | null;
    endDate: string | null;
  };
}

export function buildValidSeries(dailyHistory: DailyHistoryPayload, label: string): HistoricalPoint[] {
  const prices = dailyHistory.prices[label];
  if (!prices) return [];
  const observedIndices = dailyHistory.observedIndices?.[label];

  const points: HistoricalPoint[] = [];
  const indices = observedIndices && observedIndices.length > 0
    ? observedIndices
    : Array.from({ length: dailyHistory.dates.length }, (_, index) => index);

  for (const index of indices) {
    const value = prices[index];
    if (value === null || value === undefined || Number.isNaN(value)) continue;
    points.push({ index, date: dailyHistory.dates[index], value });
  }
  return points;
}

export function getHistoricalCoverageRange(dailyHistory: DailyHistoryPayload): {
  startDate: string | null;
  endDate: string | null;
} {
  const dates = dailyHistory.dates || [];
  return {
    startDate: dates.length > 0 ? dates[0] : null,
    endDate: dailyHistory.asOf || dates[dates.length - 1] || null,
  };
}

export function resolvePointOnOrBefore(
  dailyHistory: DailyHistoryPayload,
  label: string,
  date: string
): HistoricalPoint | null {
  const series = buildValidSeries(dailyHistory, label);
  for (let index = series.length - 1; index >= 0; index -= 1) {
    if (series[index].date <= date) {
      return series[index];
    }
  }
  return null;
}

export function resolveAnchorDateOnOrBefore(
  dailyHistory: DailyHistoryPayload,
  date: string
): string | null {
  const point = resolvePointOnOrBefore(dailyHistory, TRIGGER_ASSET, date);
  return point?.date || null;
}

export function getHistoricalPriceForDate(
  dailyHistory: DailyHistoryPayload,
  label: string,
  date: string
): { date: string; value: number } | null {
  const point = resolvePointOnOrBefore(dailyHistory, label, date);
  if (!point) return null;
  return { date: point.date, value: point.value };
}

export function getTriggerPriceForDate(
  dailyHistory: DailyHistoryPayload,
  date: string
): { date: string; value: number } | null {
  return getHistoricalPriceForDate(dailyHistory, TRIGGER_ASSET, date);
}

export function computeCustomEventReturns(
  dailyHistory: DailyHistoryPayload,
  assetMeta: Record<string, AssetMeta>,
  selectedDate: string
): CustomEventComputation {
  const coverage = getHistoricalCoverageRange(dailyHistory);
  const resolvedAnchorDate = resolveAnchorDateOnOrBefore(dailyHistory, selectedDate);

  if (!resolvedAnchorDate) {
    return {
      returnsByAsset: {},
      selectedDate,
      resolvedAnchorDate: null,
      coverage,
    };
  }

  const returnsByAsset: Record<string, Record<number, number>> = {};

  for (const label of Object.keys(assetMeta)) {
    const meta = assetMeta[label];
    const series = buildValidSeries(dailyHistory, label);
    if (series.length === 0) continue;

    let day0Index = -1;
    for (let index = series.length - 1; index >= 0; index -= 1) {
      if (series[index].date <= resolvedAnchorDate) {
        day0Index = index;
        break;
      }
    }
    if (day0Index < 0) continue;

    const denominator = series[Math.max(0, day0Index - 1)]?.value ?? series[day0Index]?.value;
    if (!denominator) continue;

    const windowStart = Math.max(0, day0Index - PRE_WINDOW_TD - 5);
    const windowEnd = Math.min(series.length, day0Index + POST_WINDOW_TD + 6);
    const window = series.slice(windowStart, windowEnd);
    const result: Record<number, number> = {};

    for (let index = 0; index < window.length; index += 1) {
      const offset = windowStart + index - day0Index;
      const value = window[index].value;
      let ret: number;
      if (meta.is_rates_bp) {
        ret = (value - denominator) * 100;
      } else {
        ret = (value / denominator - 1) * 100;
        if (!meta.invert) ret = -ret;
      }
      result[offset] = Math.round(ret * 10000) / 10000;
    }

    returnsByAsset[label] = result;
  }

  return {
    returnsByAsset,
    selectedDate,
    resolvedAnchorDate,
    coverage,
  };
}

File: engine/custom_events.py
from __future__ import annotations

from config.defaults import POST_WINDOW_TD, PRE_WINDOW_TD, TRIGGER_ASSET
from engine.models import AssetMeta, DailyHistoryPayload


def build_valid_series(daily_history: DailyHistoryPayload, label: str) -> list[tuple[int, str, float]]:
    prices = daily_history.prices.get(label)
    if not prices:
        return []
    indices = daily_history.observed_indices.get(label) or list(range(len(daily_history.dates)))
    points: list[tuple[int, str, float]] = []
    for index in indices:
        value = prices[index]
        if value is None:
            continue
        points.append((index, daily_history.dates[index], float(value)))
    return points


def get_historical_coverage_range(daily_history: DailyHistoryPayload) -> dict[str, str | None]:
    return {
        "start_date": daily_history.dates[0] if daily_history.dates else None,
        "end_date": daily_history.as_of or (daily_history.dates[-1] if daily_history.dates else None),
    }


def resolve_point_on_or_before(daily_history: DailyHistoryPayload, label: str, date: str) -> tuple[int, str, float] | None:
    series = build_valid_series(daily_history, label)
    for point in reversed(series):
        if point[1] <= date:
            return point
    return None


def resolve_anchor_date_on_or_before(daily_history: DailyHistoryPayload, date: str) -> str | None:
    point = resolve_point_on_or_before(daily_history, TRIGGER_ASSET, date)
    return point[1] if point else None


def get_historical_price_for_date(daily_history: DailyHistoryPayload, label: str, date: str) -> dict | None:
    point = resolve_point_on_or_before(daily_history, label, date)
    return None if not point else {"date": point[1], "value": point[2]}


def get_trigger_price_for_date(daily_history: DailyHistoryPayload, date: str) -> dict | None:
    return get_historical_price_for_date(daily_history, TRIGGER_ASSET, date)


def compute_custom_event_returns(
    daily_history: DailyHistoryPayload,
    asset_meta: dict[str, AssetMeta],
    selected_date: str,
) -> dict:
    coverage = get_historical_coverage_range(daily_history)
    resolved_anchor_date = resolve_anchor_date_on_or_before(daily_history, selected_date)
    if not resolved_anchor_date:
        return {
            "returns_by_asset": {},
            "selected_date": selected_date,
            "resolved_anchor_date": None,
            "coverage": coverage,
        }
    returns_by_asset: dict[str, dict[int, float]] = {}
    for label, meta in asset_meta.items():
        series = build_valid_series(daily_history, label)
        if not series:
            continue
        day0_index = -1
        for idx in range(len(series) - 1, -1, -1):
            if series[idx][1] <= resolved_anchor_date:
                day0_index = idx
                break
        if day0_index < 0:
            continue
        denominator = series[max(0, day0_index - 1)][2] if series else None
        if not denominator:
            continue
        window_start = max(0, day0_index - PRE_WINDOW_TD - 5)
        window_end = min(len(series), day0_index + POST_WINDOW_TD + 6)
        result: dict[int, float] = {}
        for pos, (_, _, value) in enumerate(series[window_start:window_end], start=window_start):
            offset = pos - day0_index
            if meta.is_rates_bp:
                ret = (value - denominator) * 100
            else:
                ret = (value / denominator - 1) * 100
                if not meta.invert:
                    ret = -ret
            result[offset] = round(ret, 4)
        returns_by_asset[label] = result
    return {
        "returns_by_asset": returns_by_asset,
        "selected_date": selected_date,
        "resolved_anchor_date": resolved_anchor_date,
        "coverage": coverage,
    }
